### PR TITLE
*: use AtomicU64 in value

### PR DESCRIPTION
**Pull request by overvenus**
Atomic value:

```
test counter::bench_counter_no_labels                   ... bench:          14 ns/iter (+/- 0)
test counter::bench_counter_with_label_values           ... bench:          98 ns/iter (+/- 7)
test counter::bench_counter_with_mapped_labels          ... bench:         367 ns/iter (+/- 53)
test counter::bench_counter_with_prepared_mapped_labels ... bench:         213 ns/iter (+/- 18)
test gauge::bench_gauge_no_labels                       ... bench:          14 ns/iter (+/- 0)
test gauge::bench_gauge_with_label_values               ... bench:          94 ns/iter (+/- 7)
```

Mutex value:

```
test counter::bench_counter_no_labels                   ... bench:          44 ns/iter (+/- 2)
test counter::bench_counter_with_label_values           ... bench:         136 ns/iter (+/- 4)
test counter::bench_counter_with_mapped_labels          ... bench:         400 ns/iter (+/- 44)
test counter::bench_counter_with_prepared_mapped_labels ... bench:         250 ns/iter (+/- 4)
test gauge::bench_gauge_no_labels                       ... bench:          43 ns/iter (+/- 1)
test gauge::bench_gauge_with_label_values               ... bench:         131 ns/iter (+/- 7)
```

cc #59

### CODE REVIEW

**Comment by siddontang** [comment]
I prefer extracting these to a AtomicFloat structure.

File: .travis.yml [hunk 1]
```diff
@@ -7,12 +7,12 @@ rust:
...
   - rust: nightly
-    env:
-      - FEATURE="--features dev"
-      - RUSTFLAGS=-Dwarnings
-
-script:
-  - cargo test $FEATURE
+    script:
+      - cargo test --features dev
+      - cargo test --features="dev nightly"
```
Comment (siddontang) [inline .travis.yml]: why two cargo test?

File: src/lib.rs [hunk 2]
```diff
@@ -41,6 +42,8 @@ mod registry;
 mod vec;
 mod histogram;
 mod push;
+#[cfg(feature="nightly")]
+mod atomicf64;
```
Comment (siddontang) [inline src/lib.rs]: where is atomicf64?

Comment (overvenus) [inline src/lib.rs]: Oops, my fault, now it's here, please review.

File: .travis.yml (hunk 1 above)
Comment (overvenus) [inline .travis.yml]: I want to test nigltly feature too.

File: src/value.rs [hunk 3]
```diff
@@ -45,10 +52,57 @@ impl ValueType {
 /// `Counter`, `Gauge`, and `Untyped`.
 pub struct Value {
     pub desc: Desc,
-    // TODO: like prometheus client go, use atomic u64.
-    pub val: RwLock<f64>,
     pub val_type: ValueType,
     pub label_pairs: Vec<LabelPair>,
+
+    #[cfg(not(feature = "nightly"))]
+    pub val: RwLock<f64>,
```
Comment (siddontang) [inline src/value.rs]: wrap all in AtomicF64, not in value.
Histogram will use AtomicF64 too. It is not good to use RwLock and AtomicF64 in two places and we have too many nightly cfg flags.

File: .travis.yml (hunk 1 above)
Comment (siddontang) [inline .travis.yml]: so why not test "dev nightly" directly?

Comment (overvenus) [inline .travis.yml]: The default feature uses `Mutex<f64>`, the nightly uses `AtomicF64`.

File: src/atomicf64.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    #[inline]
+    pub fn set(&self, val: f64) {
+        self.inner.store(val.as_u64(), Ordering::Release)
+    }
+
+    #[inline]
+    pub fn inc_by(&self, delta: f64) {
+        loop {
+            let current = self.inner.load(Ordering::Acquire);
+            let new = current.as_f64() + delta;
```
Comment (BusyJay) [inline src/atomicf64.rs]: Why not just use `std::mem::transmute`?

File: src/atomicf64.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+#[cfg(test)]
+mod test {
+    use std::f64::consts::PI;
+    use std::f64::{self, EPSILON};
+
+    use super::*;
+
+    #[cfg(feature = "nightly")]
+    #[test]
+    fn test_atomic_atomicf64() {
```
Comment (BusyJay) [inline src/atomicf64.rs]: I think it's ok to use the same test case no matter features is enable or not.

**Comment by siddontang** [comment]
Histogram can use AtomicF64 too.

File: src/atomicf64.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[cfg(not(feature = "nightly"))]
```
Comment (BusyJay) [inline src/atomicf64.rs]: Why not put these into a submodule and then reexport it like:

```
#[cfg(feature = "nightly")]
mod imp {
...
}

#[cfg(not(feature = "nightly"))]
mod imp {
...
}

pub use imp::AtomicF64;
```

Comment (siddontang) [inline src/atomicf64.rs]: +1

**Comment by overvenus** [comment]
@siddontang 
I will send a PR for histogram separately.

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay @disksing

**Comment by siddontang** [comment]
It is ok to use AtomicF64 in histogram here.

File: src/atomicf64.rs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    impl AtomicF64 {
+        pub fn new(val: f64) -> AtomicF64 {
+            AtomicF64 { inner: AtomicU64::new(val.as_u64()) }
+        }
+
+        #[inline]
+        pub fn get(&self) -> f64 {
+            self.inner.load(Ordering::Acquire).as_f64()
```
Comment (ngaut) [inline src/atomicf64.rs]: Ordering::Relex is enough

File: src/atomicf64.rs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    impl RwlockF64 {
+        pub fn new(val: f64) -> RwlockF64 {
+            RwlockF64 { inner: RwLock::new(val) }
+        }
+
+        #[inline]
+        pub fn set(&self, val: f64) {
+            *(self.inner).write().unwrap() = val;
```
Comment (BusyJay) [inline src/atomicf64.rs]: Why parentheses?

File: src/atomicf64.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                let swapped = self.inner
+                    .compare_and_swap(current.as_u64(), new.as_u64(), Ordering::Release);
+                if swapped == current {
+                    return;
+                }
+            }
+        }
+    }
+
+    trait Transform64bits: Copy {
```
Comment (BusyJay) [inline src/atomicf64.rs]: Why define a trait? Isn't it enough to just use the transmute function?

Comment (overvenus) [inline src/atomicf64.rs]: Too many unsafe code block, although they are the same. I think `as_u64` and `as_f64` are handy.

File: src/atomicf64.rs (hunk 8 above)
Comment (overvenus) [inline src/atomicf64.rs]: Better readability.

File: src/atomicf64.rs (hunk 9 above)
Comment (BusyJay) [inline src/atomicf64.rs]: I think you have more lines of code than it needs to when just using transmute. If you don't want to unsafe everywhere, just define two wrapper functions should the work too.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**